crates/proof-of-sql/src/proof_primitive/hyrax/base/hyrax_scalar.rs

```diff
@@ -0,0 +1,199 @@
+use crate::base::{
+    encode::VarInt,
+    scalar::{Scalar, ScalarConversionError},
+};
+use alloc::string::String;
+use ark_ff::{One, UniformRand};
+use ark_serialize::CanonicalSerialize;
+use core::ops::{Mul, MulAssign};
+use derive_more::{Add, AddAssign, Display, Neg, Product, Sub, SubAssign, Sum};
+use num_bigint::BigInt;
+use num_traits::{Inv, Zero};
+use serde::{Deserialize, Serialize};
+
+pub trait HyraxScalar:
+    for<'a> core::convert::From<&'a [u64; 4]> + Serialize + Scalar + Mul<Self, Output = Self>
+{
+}
+
+#[derive(
+    Serialize,
+    Deserialize,
+    Clone,
+    Copy,
+    Debug,
+    CanonicalSerialize,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Eq,
+    Default,
+    Neg,
+    AddAssign,
+    Sub,
+    SubAssign,
+    Add,
+    Display,
+    Sum,
+    Product,
+)]
+#[serde(bound = "S: for<'a> Deserialize<'a>")]
+pub struct HyraxScalarWrapper<S: HyraxScalar>(pub S);
+
+impl<S: HyraxScalar> Scalar for HyraxScalarWrapper<S> {
+    const MAX_SIGNED: Self = Self(S::MAX_SIGNED);
+
+    const ZERO: Self = Self(S::ZERO);
+
+    const ONE: Self = Self(S::ONE);
+
+    const TWO: Self = Self(S::TWO);
+
+    const TEN: Self = Self(S::TEN);
+}
+
+impl<S: HyraxScalar> TryFrom<BigInt> for HyraxScalarWrapper<S> {
+    type Error = ScalarConversionError;
+
+    fn try_from(value: BigInt) -> Result<Self, Self::Error> {
+        value.try_into().map(Self)
+    }
+}
+
+impl<S: HyraxScalar> VarInt for HyraxScalarWrapper<S> {
+    fn required_space(self) -> usize {
+        self.0.required_space()
+    }
+
+    fn decode_var(src: &[u8]) -> Option<(Self, usize)> {
+        S::decode_var(src).map(|(s, u)| (Self(s), u))
+    }
+
+    fn encode_var(self, src: &mut [u8]) -> usize {
+        S::encode_var(self.0, src)
+    }
```

Comment:
It would be nice to have a round trip test for encoding/decoding, this might help avoid headaches later.

Reply:
I think it makes more sense to put those tests on the specific Scalar (MontScalar for example), rather than HyraxScalarWrapper. If it's worthwhile, I could add some tests to verify that the wrapper implementation and the inner implementation are doing the same thing, but I don't think that would involve any round trip tests for this particular case.